File: 107. Binary Tree Level Order Traversal II.go
func levelOrderBottom(root *TreeNode) [][]int {
	var res [][]int
	if root == nil {
		return res
	}

	q := []*TreeNode{root}

	for len(q) > 0 {
		var l []int
		for i := len(q); i > 0; i-- {
			node := q[0]
			q = q[1:]
			l = append(l, node.Val)
			if node.Left != nil {
				q = append(q, node.Left)
			}
			if node.Right != nil {
				q = append(q, node.Right)
			}
		}
		res = append(res, l)
	}
	for i, j := 0, len(res)-1; i < j; i, j = i+1, j-1 {
		res[i], res[j] = res[j], res[i]
	}
	return res
}

File: 107. Binary Tree Level Order Traversal II.java
import java.util.ArrayList;
import java.util.Collections;
import java.util.LinkedList;

public class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode() {}
    TreeNode(int val) { this.val = val; }
    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}


class Solution {
    public List<List<Integer>> levelOrderBottom(TreeNode root) {
        Queue<TreeNode> q = new LinkedList<>();
        List<List<Integer>> res = new ArrayList<>();

        q.add(root);

        while(!q.isEmpty()){
            int n = q.size();
            List<Integer> l = new ArrayList<>();

            for(int i = 0; i < n; ++i){
                TreeNode node = q.poll();
                l.add(node.val);
                if(node.left != null) q.add(node.left);
                if(node.right != null) q.add(node.right);
            }
            res.add(l);
        }
        Collections.reverse(res);
        return res;
    }
}